### PR TITLE
Dump CA certificates for custom cloud

### PR DESCRIPTION
**Pull request by AbelHu**
Get CA certificates from 'http://168.63.129.16/machine?comp=acmspackage&type=cacertificates&ext=json' and save certificates to `C:\ca` in Windows nodes. Windows pods can mount this folder and install CA certificates inside the containers.

### CODE REVIEW

File: parts/windows/kuberneteswindowsfunctions.ps1 [hunk 1]
```diff
@@ -345,3 +345,41 @@ function Check-APIServerConnectivity {
...
+        if ([string]::IsNullOrEmpty($caCerts)) {
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_EMPTY_CA_CERTIFICATES -ErrorMessage "Empty after converting CA certificates rawdata to json"
+        }
+
+        $certificates = $caCerts.Certificates
+        for ($index = 0; $index -lt $certificates.Length ; $index++) {
+            $name=$certificates[$index].Name
+            $certFilePath = Join-Path $caFolder $name
+            Write-Log "Write certificate $name to $certFilePath"
+            $certificates[$index].CertBody > $certFilePath
```
Comment (mainred) [inline parts/windows/kuberneteswindowsfunctions.ps1:378]: That will be great if we could have an output example of `http://168.63.129.16/machine?comp=acmspackage&type=cacertificates&ext=json`.

File: parts/windows/kuberneteswindowsfunctions.ps1 [hunk 2]
```diff
@@ -345,3 +345,41 @@ function Check-APIServerConnectivity {
...
+        $uri = 'http://168.63.129.16/machine?comp=acmspackage&type=cacertificates&ext=json'
+
+        if (-Not (Test-Path $caFolder)) {
+            Write-Log "Create folder $caFolder for storing CA certificates"
+            New-Item -ItemType Directory -Path $caFolder
+        }
+
+        Write-Log "Download CA certificates rawdata"
+        # This is required when the root CA certs are different for some clouds.
+        $rawData = Retry-Command -Command 'Invoke-WebRequest' -Args @{Uri=$uri; UseBasicParsing=$true} -Retries 5 -RetryDelaySeconds 10
```
[the remark this answers is not in the document]
Comment (mainred) [reply parts/windows/kuberneteswindowsfunctions.ps1:362]: we are not downloading a file.

[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:362]: CA is not cached in the VHD so we need to download it in provisioning the VM

File: parts/windows/kuberneteswindowsfunctions.ps1 [hunk 3]
```diff
@@ -345,3 +345,41 @@ function Check-APIServerConnectivity {
...
+        if ([string]::IsNullOrEmpty($rawData)) {
+            Set-ExitCode -ExitCode $global:WINDOWS_CSE_ERROR_DOWNLOAD_CA_CERTIFICATES -ErrorMessage "Failed to download CA certificates rawdata"
+        }
+
+        Write-Log "Convert CA certificates rawdata"
+        $caCerts=($rawData.Content) | ConvertFrom-Json
```
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:368]: ConvertFrom-Json is a PowerShell command

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 1 above)
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:378]: it is easy to get that in any windows VM and it is too long so I think that we do not need it as a comment

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:362]: And we can handle the download failure and assign an error code. We do not need to download it to a local file and then read it for converting to json

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:368]: Converts a JSON-formatted string to a custom object or a hash table.

[the remark this answers is not in the document]
Comment (mainred) [reply parts/windows/kuberneteswindowsfunctions.ps1:368]: Thanks for your explanation.
What confused me is with the following log `"Empty after converting CA certificates rawdata to json"`, it seems we are trying to convert to a json.

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 1 above)
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:378]: Below is an example and this is one time effort so I will not leave it as a comment.
```
{
    "Version": 1,
    "Certificates": [
        {
            "Name": "Microsoft_Azure_TLS_Issuing_CA_01.cer",
            "Target": 0,
            "CertBody": "-----BEGIN CERTIFICATE-----\r\n.......\r\n-----END CERTIFICATE-----\r\n",
            "CertType": 1,
            "OCSPUrls": [
            ],
            "EKUs": [
            ],
            "Action": 0
        },
    ]
}
```

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:368]: Make sense. I will update the comment.

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 1 above)
Comment (mainred) [reply parts/windows/kuberneteswindowsfunctions.ps1:378]: Thanks, I think other reviewers will also benefit from this.

File: parts/windows/kuberneteswindowsfunctions.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (AbelHu) [reply parts/windows/kuberneteswindowsfunctions.ps1:368]: Updated the error message to "CA certificates rawdata is empty"

**Review by mainred — APPROVED** [review]
Looks good.
I think we'll also have a doc somewhere to describe the CA source and location.

**Review by xizhamsft — APPROVED** [review]

**Pull request merged**